### PR TITLE
nanopb: 0.4.6 -> 0.4.8, cross-compilation support

### PR DESCRIPTION
**Pull request by liarokapisv**
## SUMMARY

This commit:
- Bumps the nanopb version
- Adds all runtime configuration options
- Implements proper cross-compilation support which is the main use-case of the library.
- Uses newer `finalAttrs` form of `mkDerivation` to allow for easier attribute overrides.

The cross-compilation support is achieved by splitting the package into two sub-packages
consisting of the build-time generator and the runtime library.

Nanopb explicitely supports this by providing specialized `GENERATOR` and
`RUNTIME` CMake configuration options.

The top-level package uses `propagatedNativeBuildInputs` and `propagatedBuildInputs` to propagate
the sub-modules and also adds convenient symlinks to make certain uses easier.

## GENERATOR

The generator is a mostly ready-to-be-packaged python module tree.
We patch the library to also include the missing `__init__.py` and we
also fix the `PYTHON_INSTDIR` variable to follow best practice and to
prevent the library from attempting to install to a global directory.

We package the python module using `buildPythonPackage` and internally
override python in order to wrap the `nanopb_generator.py` executable.

We do *not* wrap `nanob_generator.py` due to it also being imported directly from python when
used through `protoc-gen-nanob`.

## RUNTIME

The runtime is a simple library that consists of the common
functionality among generated headers/sources.

It is configured through `preprocessor definitions` and consumer projects *must*
be compiled with the same definitions.

This is currently achieved by exposing all configuration options through the top-level overrides and
patching the CMakeLists.txt so that the definitions are added to the to-be-installed
CMake targets as PUBLIC properties.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
  - [x] Also checked cross-compilation to `arm-none-eabi(hf)` targets with the above.
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Thesola10** [comment]
Are the edits to `maintainers-list.nix` on purpose? They should probably be broken off into a separate PR

**Comment by liarokapisv** [comment]
https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions  implies that one can add themselves to the maintainers list in the same PR as long as it is done in a different commit.

**Comment by liarokapisv** [comment]
I want to note that I did/have issues with the tests of the criterion package. The build works properly but the tests fail and I am not yet sure why that is.  I had to locally override the version to 0.4.6 in order for them to work otherwise they all fail.

**Comment by liarokapisv** [comment]
I went through the process of updating `nanopb` to the latest version (`0.4.8`). This version removes the need for our patches. I also integrated the fixes for criterion's tests, so now `nixpkgs-review` passes without any overrides.

**Comment by liarokapisv** [comment]
The latest push patches the CMakeLists.txt so that the configuration options are added as public compile definition properties to the CMake targets.

File: pkgs/by-name/na/nanopb/generator-out.nix [hunk 1]
```diff
@@ -0,0 +1,59 @@
+{ stdenv
+, cmake
+, python3
+, writeTextFile
+, protobuf
+, src
+, version
+}:
+let
+  pyproject_toml = writeTextFile {
```
Comment (Thesola10) [inline pkgs/by-name/na/nanopb/generator-out.nix:10]: Have you considered contributing this `pyproject.toml` upstream? Apparently they used to have one so I think they'd be open to it.

**Review by Thesola10 — APPROVED** [review]
LGTM, works with my criterion project as well, doesn't implode

File: pkgs/by-name/na/nanopb/generator-out.nix (hunk 1 above)
Comment (liarokapisv) [reply pkgs/by-name/na/nanopb/generator-out.nix:10]: Yea I have a lot going on currently but I am planning to do so. I also want to add the cmake options to the targets as well, it is a much better way to handle their propagation than relying on package managers to handle it.

**Review by Yumasi — APPROVED** [review]

File: pkgs/by-name/na/nanopb/package.nix [hunk 2]
```diff
@@ -0,0 +1,122 @@
+{ stdenvNoCC
+, callPackage
+, fetchFromGitHub
+, buildPackages
+, lib
+, enable_malloc ? false
```
Comment (Mindavi) [inline pkgs/by-name/na/nanopb/package.nix:6]: I think the convention is to use camelCase, not snake_case, for package options.

**Comment by Mindavi** [comment]
Other than my one comment, I think this looks pretty good. But do follow conventions as much as possible :)

**Review by otavio — APPROVED** [review]

**Review by Thesola10 — APPROVED** [review]

**Review by kalbasit — APPROVED** [review]
Nice work @liarokapisv thank you for working on this!

**Comment by kalbasit** [comment]
Result of `nixpkgs-review pr 288422` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>criterion</li>
    <li>criterion.dev</li>
    <li>nanopb</li>
    <li>nanopbMalloc</li>
  </ul>
</details>

**Comment by kalbasit** [comment]
Unless anyone objects, I will merge this later today. I'm waiting for the eval to be completed before I do so.

**Comment by kalbasit** [comment]
backport failed because #315684 did not get backported, I've opened #322985 to get that backported so I can retry this one and get it to work.

**Comment by bouk** [comment]
I'm not able to run the nanopb generator, getting this error:

```
~/s/g/t/mech (master) nanopb_generator
Traceback (most recent call last):
  File "/nix/store/721nzxig9ybsisg28vkxaysqr2p4051r-nanopb-generator-0.4.8/bin/.nanopb_generator-wrapped", line 6, in <module>
    from nanopb_generator import *
  File "/nix/store/721nzxig9ybsisg28vkxaysqr2p4051r-nanopb-generator-0.4.8/bin/nanopb_generator.py", line 22, in <module>
    from nanopb.generator.nanopb_generator import main_cli, main_plugin
ModuleNotFoundError: No module named 'nanopb'
```

**Pull request merged**